File: frontend/src/api/craCompliance.ts
import apiClient from './client'

export interface CraRequirementResult {
  id: string
  assessment_id: string
  requirement_id: string
  requirement_title: string
  annex_part: number
  status: string
  auto_populated: boolean
  evidence_summary: string | null
  finding_ids: string[]
  tool_sources: string[]
  manual_notes: string | null
  manual_evidence: string | null
  related_cwes: string[]
  related_cves: string[]
  assessed_at: string | null
  updated_at: string
}

export interface CraAssessment {
  id: string
  project_id: string
  firmware_id: string | null
  assessor_name: string | null
  product_name: string | null
  product_version: string | null
  overall_status: string
  auto_pass_count: number
  auto_fail_count: number
  manual_count: number
  not_tested_count: number
  created_at: string
  updated_at: string
  requirement_results: CraRequirementResult[]
}

export type CraAssessmentSummary = Omit<CraAssessment, 'requirement_results'>

export async function createCraAssessment(projectId: string, body?: {
  product_name?: string; product_version?: string; assessor_name?: string; firmware_id?: string
}): Promise<CraAssessment> {
  const { data } = await apiClient.post<CraAssessment>(`/projects/${projectId}/cra/assessments`, body || {})
  return data
}

export async function listCraAssessments(projectId: string): Promise<CraAssessmentSummary[]> {
  const { data } = await apiClient.get<CraAssessmentSummary[]>(`/projects/${projectId}/cra/assessments`)
  return data
}

export async function getCraAssessment(projectId: string, assessmentId: string): Promise<CraAssessment> {
  const { data } = await apiClient.get<CraAssessment>(`/projects/${projectId}/cra/assessments/${assessmentId}`)
  return data
}

export async function autoPopulateCra(projectId: string, assessmentId: string): Promise<CraAssessment> {
  const { data } = await apiClient.post<CraAssessment>(`/projects/${projectId}/cra/assessments/${assessmentId}/auto-populate`)
  return data
}

export async function updateCraRequirement(
  projectId: string, assessmentId: string, requirementId: string,
  body: { status?: string; manual_notes?: string; manual_evidence?: string }
): Promise<CraRequirementResult> {
  const { data } = await apiClient.patch<CraRequirementResult>(
    `/projects/${projectId}/cra/assessments/${assessmentId}/requirements/${requirementId}`, body
  )
  return data
}

export async function exportCraChecklist(projectId: string, assessmentId: string): Promise<Record<string, unknown>> {
  const { data } = await apiClient.get<Record<string, unknown>>(`/projects/${projectId}/cra/assessments/${assessmentId}/export`)
  return data
}

export async function exportArticle14(projectId: string, assessmentId: string, cveId: string): Promise<Record<string, unknown>> {
  const { data } = await apiClient.get<Record<string, unknown>>(`/projects/${projectId}/cra/assessments/${assessmentId}/article14/${cveId}`)
  return data
}

File: backend/app/models/cra_compliance.py
import uuid
from datetime import datetime

from sqlalchemy import Boolean, ForeignKey, Integer, String, Text, func
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class CraAssessment(Base):
    __tablename__ = "cra_assessments"

    id: Mapped[uuid.UUID] = mapped_column(
        primary_key=True,
        default=uuid.uuid4,
        server_default=func.gen_random_uuid(),
    )
    project_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("projects.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    firmware_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("firmware.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    assessor_name: Mapped[str | None] = mapped_column(String(255))
    product_name: Mapped[str | None] = mapped_column(String(255))
    product_version: Mapped[str | None] = mapped_column(String(100))
    overall_status: Mapped[str] = mapped_column(
        String(20), default="in_progress", server_default="in_progress"
    )
    auto_pass_count: Mapped[int] = mapped_column(
        Integer, default=0, server_default="0"
    )
    auto_fail_count: Mapped[int] = mapped_column(
        Integer, default=0, server_default="0"
    )
    manual_count: Mapped[int] = mapped_column(
        Integer, default=0, server_default="0"
    )
    not_tested_count: Mapped[int] = mapped_column(
        Integer, default=0, server_default="0"
    )
    created_at: Mapped[datetime] = mapped_column(server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        server_default=func.now(), onupdate=func.now()
    )

    requirement_results: Mapped[list["CraRequirementResult"]] = relationship(
        back_populates="assessment", cascade="all, delete-orphan"
    )
    project: Mapped["Project"] = relationship()  # noqa: F821


class CraRequirementResult(Base):
    __tablename__ = "cra_requirement_results"

    id: Mapped[uuid.UUID] = mapped_column(
        primary_key=True,
        default=uuid.uuid4,
        server_default=func.gen_random_uuid(),
    )
    assessment_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("cra_assessments.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    requirement_id: Mapped[str] = mapped_column(String(50), nullable=False)
    requirement_title: Mapped[str] = mapped_column(String(255), nullable=False)
    annex_part: Mapped[int] = mapped_column(Integer, nullable=False)
    status: Mapped[str] = mapped_column(
        String(20), default="not_tested", server_default="not_tested"
    )
    auto_populated: Mapped[bool] = mapped_column(
        Boolean, default=False, server_default="false"
    )
    evidence_summary: Mapped[str | None] = mapped_column(Text)
    finding_ids: Mapped[dict] = mapped_column(
        JSONB, server_default="[]"
    )
    tool_sources: Mapped[dict] = mapped_column(
        JSONB, server_default="[]"
    )
    manual_notes: Mapped[str | None] = mapped_column(Text)
    manual_evidence: Mapped[str | None] = mapped_column(Text)
    related_cwes: Mapped[dict] = mapped_column(
        JSONB, server_default="[]"
    )
    related_cves: Mapped[dict] = mapped_column(
        JSONB, server_default="[]"
    )
    assessed_at: Mapped[datetime | None] = mapped_column(nullable=True)
    updated_at: Mapped[datetime] = mapped_column(
        server_default=func.now(), onupdate=func.now()
    )

    assessment: Mapped["CraAssessment"] = relationship(
        back_populates="requirement_results"
    )
